Fix bandmap

dl1jbe opened the pull request:
* Fix memory leak during grab of station from bandmap
* Fix segfault during bandmap_show() and grab_next() under heavy load.
  (tnx DH5FS for reporting)

zcsahok commented:
Will check later today.

I've seen this assertion error:
```g_ascii_strdown: assertion 'str != NULL'```
Could this be also related? (I had dupe display on)

dl1jbe commented:
Am Sat, 06 Feb 2021 04:13:13 -0800
schrieb Zoltan Csahok <notifications@github.com>:

> Will check later today.
> 
Fine. Saw you in contest (and also Ervin) in the morning, so take your
time.

> I've seen this assertion error:
> ```g_ascii_strdown: assertion 'str != NULL'```
> Could this be also related? (I had dupe display on)

Yes, there are some more but all for the same reason.

We filter out some spots (by band, mode, ...) from the 'allspots' list
to the 'spots' array and use the latter for grab, bandmap_show and
others.

If a new spot comes in with a frequency already occupied by a former
call it may happen, that the former spot is in 'spots' array but gets
replaced by the new spot in the list. As we had only references to the
spots from the list in the array that leads to a segfault whenever we
try to use the entry.

I changed it to make real copies of the old spot into 'spots' array. So
whatever happens to the 'allspots' list will not influence any spot in
the array after the filtering has happened.

Seems to work without problems.

73, de Tom



-- 
"Do what is needful!"
Ursula LeGuin: Earthsea
--

airween commented:
Just my 2 cents to the affected [tlf_banmap issue](https://github.com/zcsahok/tlf_bandmap/issues/1): in first release it used the `$HOME/.bmdata.dat`, but Tlf stores it in CWD (I saw that's changed now).

If you touch the related code, we can move the `$CWD/.bmdata.dat` to `$HOME`, and can rename it to `.tlf_bmdata.dat`. 

Thoughts?

zcsahok reviewed (commented):
Added some points.

Spot handling from `bm_add()` up is OK, but the things happening from socket reading until `bm_add()` are not in a good shape. But at least they work...

Some random things from this area (unrelated to this PR):
- the call `i = recvline(&prsock, line, BUFFERSIZE - 1);` in `splitscreen.c` (the names contain history but do not reflect the actual actions) overwrites the value of `prsock`. Luckily it is not actually used.
    - re: splitscreen: maybe we shall suppress displaying spots on non-visible screens to avoid the see-through effect.
- `addlog()` in the same file seemingly expects a potentially multi-line buffer as it has a for-loop to process it. Later in the function `bm_add()` works on a single line. Maybe we lose some of spots here? Not sure.
- the juggling with `spot_ptr[]` (which is not a "real" pointer) and `spotarray[]` (which *is* pointers/indices) could be also reviewed/simplified
- `sockserv.c` can do literally `everything` (not sure if all correctly, though) but we need just a tiny part. `netkeyer` doesn't even use it.

zcsahok commented:
@airween re: bmdata.dat location: the current dir is OK, as this allows running multiple instances for the same user. tlf_bandmap can take the file from any location now.

Using bmdata.dat was just to check the feasibility. It we stay with it then we could extend its header line with current freq and other info.
For point-and-clink grabbing the existing UDP networking could be extended with such a method.
Just brainstorming.

airween commented:
> @airween re: bmdata.dat location: the current dir is OK, as this allows running multiple instances for the same user.

right, I didn't thought about multiple instances.

> For point-and-clink grabbing the existing UDP networking could be extended with such a method.

I had same idea about that :).

I can make this extension soon in next few days in Tlf (today I'll try to participate on EUDX).

dl1jbe commented on src/bandmap.c:
```diff
@@ -81,8 +81,9 @@ bm_config_t bm_config = {
+spot *copy_spot(spot *data);
 
 /*
  * write bandmap spots to a file
```
No Idea why it is here. See commit 2a05cd660049b40e6d907dfdb0785c85be917e7f

zcsahok replied:
@airween any idea? IMO chomping it while reading from file should be sufficient. (dxccdata->pfx is already chomped, I assume)

dl1jbe commented on src/bandmap.c:
```diff
@@ -656,6 +656,54 @@ freq_t bm_get_center(int band, int mode) {
+		(bm_config.showdupes || !dupe) &&
+		(! bm_config.onlymults || multi)) {
+
+	    data -> dupe = dupe;
```
Iirc idea was that the dupe state is only needed for display, grab and so on. 
But yes, setting it for all would enable us to skip the check next time we reevaluate the filter.

zcsahok replied:
I think the next check can't be avoided as the log state could have changed in the meantime. Actually, this is also true for multi info. But that one is not stored. (we have no flags to indicate that log/spots were updated and are potentially out of sync; those would help to avoid unnecessary evaluations) 

I mean more that if we don't update the dupe state of non-visible spots then on tlf_bandmap they'll have an invalid state.

dl1jbe replied:
The question is: can a dupe state later be gone? Than the actual code is already bugyy. Otherwise we need to only check for dupe stations which were NOT dupe before.

Just reading your second section: Yes it needs to be done for the external tools. Will add a fix for it - not sure if I manage to do it today.

zcsahok replied:
Dupe state can be gone in a minitest.

Let's do it a separate PR. At least we have went through the potential issues a bit.

dl1jbe commented on src/bandmap.c:
```diff
@@ -656,6 +656,54 @@ freq_t bm_get_center(int band, int mode) {
+	multi = bm_ismulti(NULL, data, data->band);
+	dupe = bm_isdupe(data->call, data->band);
+
+	if ((!iscontest || !IsWarcIndex(data->band))         &&
```
Some optimization is surely possible, but please reread the contest check - if we are in no contest we show spots also for non WARC bands.

zcsahok replied:
with "contest check" I meant
```
    if (iscontest && IsWarcIndex(data->band))  continue;
```

But if multi/dupe has to be always evaluated then we gain nothing, agree.

dl1jbe replied:
Ok - understand. Let us take it on the todo list for next days.

dl1jbe commented:
> Added some points.

Thanks. Focus atm was on fixing the bugs. But thanks for the hints to improvement.


> Spot handling from `bm_add()` up is OK, but the things happening from socket reading until `bm_add()` are not in a good shape. But at least they work...
> 

As far as I understood the sources and available history notes Rein picked up the code for splitscreen.c and sockserv.c from an existing cluster client and integrated in into TLF. As the code is really tangled I did stay away from it as much as possible. But sure in the long run we need to sort things out and decide if a full cluster interface is needed inside TLF.

> Some random things from this area (unrelated to this PR):
> 
>     * the call `i = recvline(&prsock, line, BUFFERSIZE - 1);` in `splitscreen.c` (the names contain history but do not reflect the actual actions) overwrites the value of `prsock`. Luckily it is not actually used.
>       
>       * re: splitscreen: maybe we shall suppress displaying spots on non-visible screens to avoid the see-through effect.
> 
>     * `addlog()` in the same file seemingly expects a potentially multi-line buffer as it has a for-loop to process it. Later in the function `bm_add()` works on a single line. Maybe we lose some of spots here? Not sure.
> 
Looks like the original code accumulated more than one network message before getting time to catch up with the workload. I am not sure, but it looks as if that can never happen in actual situation - difficult to think through.

>     * the juggling with `spot_ptr[]` (which is not a "real" pointer) and `spotarray[]` (which _is_ pointers/indices) could be also reviewed/simplified
> 
Yep.

>     * `sockserv.c` can do literally `everything` (not sure if all correctly, though) but we need just a tiny part. `netkeyer` doesn't even use it.

There is a further point - 'addtext' includes some logic I do not understand as I am not quite common with the old usage of dx clusters in the 90ies. It was the time when some of the traffic came in via packet. Please have a look into the function for 'private cluster messages' and 'convers mode' - I can not find a lot of sense in it. Is it still used today? If not we should clean up here.

dl1jbe commented:
> > For point-and-clink grabbing the existing UDP networking could be extended with such a method.
> 
> I had same idea about that :).
> 
> I can make this extension soon in next few days in Tlf (today I'll try to participate on EUDX).

I would make haste on it. Better sit down for a moment and think over:

* If we start in these direction there will be more ideas to make use of TLFs internal state.
* Then we should look for an extensible way to get data from TLF and maybe push something into it.  The quasi standard for doing so is to incorporate an XMLRPC server into the program - see fldigi and others.  I think that may be a better solution at least in the long run.

N0NB commented:
> Just my 2 cents to the affected [tlf_banmap issue](https://github.com/zcsahok/tlf_bandmap/issues/1): in first release it used the `$HOME/.bmdata.dat`, but Tlf stores it in CWD (I saw that's changed now).
> 
> If you touch the related code, we can move the `$CWD/.bmdata.dat` to `$HOME`, and can rename it to `.tlf_bmdata.dat`.
> 
> Thoughts?

I would prefer not polluting $HOME if possible.  As I see it, .bmdata.dat is not a config file and is really specific to a particular event, i.e. some events allow single ops to use spotting while others put them into an unlimited category, etc.  I think $CWD is the proper place for it.

dl1jbe commented:
Thanks. Lets keep the PR open for now until relevant discussion moved to a new Issue.

airween commented:
>     * Then we should look for an extensible way to get data from TLF and maybe push something into it.  The quasi standard for doing so is to incorporate an XMLRPC server into the program - see fldigi and others.  I think that may be a better solution at least in the long run.

yes, the next idea was the XMLRPC server inside of Tlf - but then we should think about which XMLRPC lib can use - "old" libxmlrpc-c, or flxmlrpc.

We can use this server for other functions too.
